pkg/namespace/namespace.go

```diff
@@ -0,0 +1,130 @@
+package namespace
+
+import (
+	"bytes"
+	"fmt"
+)
+
+type Namespace struct {
+	Version uint8
+	ID      []byte
+}
+
+// New returns a new namespace with the provided version and id.
+func New(version uint8, id []byte) (Namespace, error) {
+	err := validateVersion(version)
+	if err != nil {
+		return Namespace{}, err
+	}
+
+	err = validateID(version, id)
+	if err != nil {
+		return Namespace{}, err
+	}
```

Comment:
should this validation logic go in the validation method?

Reply:
It can but why should the `Namespace` type support creating a new namespace that violates the rules of version / ID? Then consumers need to call `validateBasic()` prior to using a Namespace. If we enforce these variants at construction time (i.e. `New()`) then users don't need to validate the namespace at each usage.

Reply:
https://github.com/celestiaorg/celestia-app/pull/1529/files#r1150771472